README
This is Perl module Kossy.

INSTALLATION

Kossy installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Kossy

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Kossy documentation is available as in POD. So you can do:

    % perldoc Kossy

to read the documentation online with your favorite pager.

Masahiro Nagano


Changes
Revision history for Perl extension Kossy

{{$NEXT}}

0.63 2023-11-13T02:24:23Z

        - Stabilize routing #22

0.62 2023-11-09T08:56:51Z

        - Fixed JSON request #21

0.61 2023-11-06T14:26:54Z

        - Enhance Kossy::Request to support array values in JSON body params #19 Thank you kfly8
        - Support 5.14 and above

0.60 2021-09-16T12:04:10Z

        - [IMCOMPATIBLE CHANGE] fix: defense from JSON hijacking (Thank you kfly8)
        - Customize json serializer (Thank you kfly8)

0.50 2021-08-26T13:49:56Z

        - add halt methods (Thank you kfly8)
        - fix cookie tests (Thank you kfly8)
        - remove BOM from json (Thank you kfly8)

0.40 2016-07-19T15:03:36Z

        - Fix multibyte path handling (Thanks songmu)

0.39 2014-10-20T05:46:07Z

        - fix some document bug (Thanks hisaichi5518, avvmoto, niku4i)
        - improve perfomance
          cache router result
          add Kossy::Header
          render json

0.38 2014-05-29T06:35:40Z

        - re-packaging with Minilla

0.37    Thu May 29 1:35:15 2014
        - fix packaging issue. META file did not updated

0.36    Thu May 29 0:43:12 2014
        - fix packaging issue. change version in all modules

0.35    Wed May 28 23:54:55 2014
        - decode path args as utf8. Router::Boom encodes path?

0.34    Wed Feb 20 2:18:18 2014
        - avoid 0 0 length query_string in uri_for

0.33    Wed Feb 20 1:12:34 2014
        - skip parsing body if content_type is null
        - requires HTTTP:Entity::Parser 0.11

0.32    Wed Feb 19 15:53:33 2014
        - use build_urlencoded that is provided by WWW::Form::UrlEncoded in uri_for

0.31    Wed Feb 12 13:51:54 2014
        - fix packaging issue. change version in all modules

0.30    Wed Feb 12 13:46:15 2014
        - replace Kossy::BodyParser with HTTP::Entity::Parser

0.28    Thu Nov 28 10:29:51 2013
        - Kossy::Request->new_response returns Kossy::Response

0.27    Tue Nov 12 14:45:09 2013
        - fix packaging issue

0.26    Tue Nov 12 11:07:05 2013
        - add support for changing X-Frame-Options through $c->res
        - document parse json body

0.25    Wed Nov 6 11:19:31 2013
        - non trial release
        - switch to use cpanfile

0.24    Thu Oct 31 11:46:31 2013
        - add Kossy::BodyParser based on tokuhiorm's code
          improve perfomance of POST request

0.23    Mon Oct 21 14:17:49 2013
        - improve performance
          use HTTP::Headers::Fast
          optimze Kossy::Response->finalize

0.22    Tue Oct 15 11:44::46 2013
        - switch to using Router::Boom

0.21    Thu Oct 10 15:37:36 2013
        - [Kossy::Request] cache base() for performance. URI->canonical is slow

0.20    Wed Oct 9 15:46:10 2013
        - separete Kossy::Validator to another dist

0.19    Fri Aug 30 16:00:56 2013
        - fixed requires

0.18    Thu Aug 8 16:05:45 2013
        - json escape only for string ( Thank you sonots-san)

0.17    Tue Jul 16 13:38:09 2013
        - updated to Twitter Bootstrap
        - updated to jQuery

0.16    Tue Jul 16 13:29:21 2013
        - added floating number validation rules

0.15    Fri Apr 0:29:46 2013
        - render_json: enable allow_blessed and convert_blessed

0.14    Thu Mar 14 22:25:49 2013
        - Kossy::Validator: support Coderef default

0.13    Wed Dec 5 11:23:43 2012
        - requires Text::Xslate >= 1.6001, HTML::FillInForm::Lite >= 1.11

0.12    Fri Aug 24 18:17:51 2012
        - updated to Twitter Bootstrap 2.1.0
        - updated to jQuery 1.8.0

0.11    Tue Jul 24 21:48:04 2012
        - fixed uri_for() if app is mounted on non '/' location. (Thank you fujiwara-san)

0.10    Tue Jun 19 11:12:47 2012
        - accept hash argument in new constructor

0.09    Thu May 17 17:48:24 2012
        - fixed css uri in skelton

0.08    Tue Apr 24 18:59:47 2012
        - updated to Twitter Bootstrap 2.0.3
        - updated to jQuery 1.7.2
        - using sharedir for assets, Kossy::Assets was deprecated

0.07    Fri Feb 17 12:07:06 2012
        - updated to Twitter Bootstrap 2.0

0.06    Thu Jan 5 10:31:51 2012
        - htmlize error page

0.05    Tue Nov 29 17:43:51 2011
        - release to CPAN

0.04    Tue Nov 29 01:04:55 2011
        - addes Assets for setup

0.03    Mon Nov 28 19:20:09 2011
        - render_json security staff.
        - added document
        - added router method for dispach other than GET and POST

0.02    Fri Nov 11 10:35:14 2011
        - added Number::Format to Xslate default functions.

0.01    Fri Nov 11 10:35:14 2011
        - original version


lib/Kossy.pm
package Kossy;

use strict;
use warnings;
use 5.014004;
use utf8;
use Carp qw//;
use Cwd qw//;
use File::Basename qw//;
use Text::Xslate;
use HTML::FillInForm::Lite qw//;
use JSON qw//;
use Scalar::Util qw//;
use Try::Tiny;
use Encode;
use Router::Boom::Method;
use Class::Accessor::Lite (
    new => 0,
    rw => [qw/root_dir/]
);
use base qw/Exporter/;
use Kossy::Exception;
use Kossy::Connection;
use Kossy::Request;
use Kossy::Response;
use HTTP::Headers::Fast;

our $VERSION = '0.63';
our @EXPORT = qw/new root_dir psgi build_app _router _connect get post router filter _wrap_filter/;

our $XSLATE_CACHE = 1;
our $XSLATE_CACHE_DIR;
our $SECURITY_HEADER = 1;
our $JSON_SERIALIZER;

# cache underscore translation
HTTP::Headers::Fast::_standardize_field_name('X-Frame-Options');

sub new {
    my $class = shift;
    my %args;
    if ( @_ < 2 ) {
        my $root_dir = shift;
        my @caller = caller;
        $root_dir ||= File::Basename::dirname( Cwd::realpath($caller[1]) );
        $args{root_dir} = $root_dir;
    }
    else {
        %args = @_;
    }

    bless \%args, $class;
}

sub psgi {
    my $self = shift;
    if ( ! ref $self ) {
        my %args;
        if ( @_ < 2 ) {
            my $root_dir = shift;
            my @caller = caller;
            $root_dir ||= File::Basename::dirname( Cwd::realpath($caller[1]) );
            $args{root_dir} = $root_dir;
        }
        else {
            %args = @_;
        }
        $self = $self->new(%args);
    }

    $self->build_app;
}

sub build_app {
    my $self = shift;

    my $security_header_local = $SECURITY_HEADER;
    my %match_cache;

    my $tx = __PACKAGE__->_build_text_xslate(
        root_dir  => $self->root_dir,
        cache     => $XSLATE_CACHE,
        cache_dir => $XSLATE_CACHE_DIR,
    );

    my $json_serializer = __PACKAGE__->_build_json_serializer();

    sub {
        my $env = shift;
        $Kossy::Response::SECURITY_HEADER = $security_header_local;
        try {
            my $header = bless {
                'content-type' => 'text/html; charset=UTF-8',
                $security_header_local ? ('x-frame-options' => 'DENY') : (),
            }, 'HTTP::Headers::Fast';
            my $c = Kossy::Connection->new({
                tx => $tx,
                req => Kossy::Request->new($env),
                res => Kossy::Response->new(200, $header),
                stash => {},
                json_serializer => $json_serializer,
            });
            my $method = uc($env->{REQUEST_METHOD});
            my $cache_key = $method . '-' . $env->{PATH_INFO};
            my ($match,$args) = try {
                if ( exists $match_cache{$cache_key} ) {
                    return @{$match_cache{$cache_key}};
                }
                my $path_info = Encode::decode_utf8( $env->{PATH_INFO},  Encode::FB_CROAK | Encode::LEAVE_SRC );
                my ($dest, $captured, $method_not_allowed) = $self->_router->match($method, $path_info);
                if ($method_not_allowed) {
                    $c->halt(405);
                }

                if (!$dest) {
                    $c->halt(404);
                }

                $match_cache{$cache_key} = [$dest, $captured];
                return ($dest, $captured);
            } catch {
                if ( ref $_ && ref $_ eq 'Kossy::Exception' ) {
                    die $_; #rethrow
                }
                $c->halt(400,'unexpected character in request');
            };

            my $code = $match->{__action__};
            my $filters = $match->{__filter__} || [];
            if ( $] == 5.020000 || $] == 5.020100 ) {
                # workaround for 5.20.0 or 5.20.1 https://github.com/kazeburo/Kossy/pull/10
                my %args = map { $_ => Encode::decode_utf8($args->{$_}) } keys %$args;
                $c->args(\%args);
            } else {
                $c->args({%$args});
            }
            my $app = sub {
                my ($self, $c) = @_;
                my $response;
                my $res = $code->($self, $c);
                Carp::croak "Undefined Response" if ! defined $res;
                my $res_t = ref($res) || '';
                if ( $res_t eq 'Kossy::Response' ) {
                    $response = $res;
                }
                elsif ( $res_t eq 'Plack::Response' ) {
                    $response = bless $res, 'Kossy::Response';
                }
                elsif ( $res_t eq 'ARRAY' ) {
                    $response = Kossy::Response->new(@$res);
                }
                elsif ( !$res_t ) {
                    $c->res->body($res);
                    $response = $c->res;
                }
                else {
                    Carp::croak sprintf "Unknown Response: %s", $res_t;
                }
                $response;
            };

            for my $filter ( reverse @$filters ) {
                $app = $self->_wrap_filter($filter,$app);
            }
            # do all
            local $Kossy::Response::DIRECT;
            $app->($self, $c)->finalize;
        } catch {
            if ( ref $_ && ref $_ eq 'Kossy::Exception' ) {
                return $_->response;
            }
            die $_;
        };
    };
}

sub _build_text_xslate {
    my ($class, %args) = @_;
    my ($root_dir, $cache, $cache_dir) = @args{qw/root_dir cache cache_dir/};

    my $fif = HTML::FillInForm::Lite->new();
    my $tx = Text::Xslate->new(
        path => [ $root_dir . '/views' ],
        input_layer => ':utf8',
        module => ['Text::Xslate::Bridge::TT2Like','Number::Format' => [':subs']],
        function => {
            fillinform => sub {
                my $q = shift;
                return sub {
                    my ($html) = @_;
                    return Text::Xslate::mark_raw( $fif->fill( \$html, $q ) );
                }
            }
        },
        cache => $cache,
        defined $cache_dir ? ( cache_dir => $cache_dir ) : (),
    );
    return $tx
}

sub _build_json_serializer {
    my $class = shift;

    if (defined $JSON_SERIALIZER) {
        if (Scalar::Util::blessed($JSON_SERIALIZER) && $JSON_SERIALIZER->can('encode')) {
            return $JSON_SERIALIZER
        }
        else {
            Carp::croak '$Kossy::JSON_SERIALIZER must have `encode` method';
        }
    }

    # default case
    return JSON->new()->allow_blessed(1)->convert_blessed(1)->ascii(1);
}


my $_ROUTER={};
sub _router {
    my $klass = shift;
    my $class = ref $klass ? ref $klass : $klass;
    if ( !$_ROUTER->{$class} ) {
        $_ROUTER->{$class} = Router::Boom::Method->new;
    }
    $_ROUTER->{$class};
}

sub _connect {
    my $class = shift;
    my ( $methods, $pattern, $filter, $code ) = @_;
    $methods = ref($methods) ? $methods : [$methods];
    if (!$code) {
        $code = $filter;
        $filter = [];
    }
    unless ( (ref $code ||'') eq 'CODE') {
        Carp::croak "\$code argument must be a CODE reference";
    }
    $class->_router->add($methods, $pattern, {
        __action__ => $code,
        __filter__ => $filter
    });
}

sub get {
    my $class = caller;
    $class->_connect( ['GET','HEAD'], @_  );
}

sub post {
    my $class = caller;
    $class->_connect( ['POST'], @_  );
}

sub router {
    my $class = caller;
    $class->_connect( @_  );
}

my $_FILTER={};
sub filter {
    my $class = caller;
    if ( !$_FILTER->{$class} ) {
        $_FILTER->{$class} = {};
    }
    if ( @_ ) {
        $_FILTER->{$class}->{$_[0]} = $_[1];
    }
    $_FILTER->{$class};
}

sub _wrap_filter {
    my $klass = shift;
    my $class = ref $klass ? ref $klass : $klass;
    if ( !$_FILTER->{$class} ) {
        $_FILTER->{$class} = {};
    }
    my ($filter,$app) = @_;
    my $filter_subref = $_FILTER->{$class}->{$filter};
    Carp::croak sprintf("Filter:%s is not exists", $filter) unless $filter_subref;
    return $filter_subref->($app);
}

1;

__END__

=head1 NAME

Kossy - Sinatra-ish Simple and Clear web application framework

=head1 SYNOPSIS

  % kossy-setup MyApp
  % cd MyApp
  % plackup app.psgi

  ## lib/MyApp/Web.pm

  use Kossy;

  get '/' => sub {
      my ( $self, $c )  = @_;
      $c->render('index.tx', { greeting => "Hello!" });
  };

  get '/json' => sub {
      my ( $self, $c )  = @_;
      my $result = $c->req->validator([
          'q' => {
              default => 'Hello',
              rule => [
                  [['CHOICE',qw/Hello Bye/],'Hello or Bye']
              ],
          }
      ]);
      $c->render_json({ greeting => $result->valid->get('q') });
  };

  1;

  ## views/index.tx
  : cascade base
  : around content -> {
    <: $greeting :>
  : }

=head1 DESCRIPTION

Kossy is Sinatra-ish Simple and Clear web application framework, which is based upon L<Plack>, L<Router::Boom>, L<Text::Xslate> and build-in Form-Validator. That's suitable for small application and rapid development.

=head1 Kossy class

Kossy exports some methods to building application

=head2 CLASS METHODS for Kossy class

=over 4

=item my $kossy = Kossy->new( root_dir => $root_dir );

Create instance of the application object.

=back

=head2 OBJECT METHODS for Kossy class

=over 4

=item my $root_dir = $kossy->root_dir();

accessor to root directory of the application

=item my $app = $kossy->psgi();

return PSGI application

=back

=head2 DISPATCHER METHODS for Kossy class

=over 4

=item filter

makes application wrapper like plack::middlewares.

  filter 'set_title' => sub {
      my $app:CODE = shift;
      sub {
          my ( $self:Kossy, $c:Kossy::Connection )  = @_;
          $c->stash->{site_name} = __PACKAGE__;
          $app->($self,$c);
      }
  };

=item get path:String => [[filters] =>] CODE

=item post path:String => [[filters] =>] CODE

setup router and dispatch code

  get '/' => [qw/set_title/] => sub {
      my ( $self:Kossy, $c:Kossy::Connection )  = @_;
      $c->render('index.tx', { greeting => "Hello!" });
  };

  get '/json' => sub {
      my ( $self:Kossy, $c:Kossy::Connection )  = @_;
      $c->render_json({ greeting => "Hello!" });
  };

dispatch code shall return Kossy::Response object or PSGI response ArrayRef or String.

=item router 'HTTP_METHOD'|['METHOD'[,'METHOD']] => path:String => [[filters] =>] CODE

adds routing rule other than GET and POST

  router 'PUT' => '/put' => sub {
      my ( $self:Kossy, $c:Kossy::Connection )  = @_;
      $c->render_json({ greeting => "Hello!" });
  };

=back

=head1 Kossy::Connection class

per-request object, herds request and response

=head2 OBJECT METHODS for Kossy::Connection class

=over 4

=item req:Kossy::Request

=item res:Kossy::Response

=item stash:HashRef

=item args:HashRef

path parameters captured by Router::Boom

    get '/user/:id' => sub {
        my ($self, $c) = @_;
        my $id = $c->args->{id};
        ...
    };

=item halt(status_code, message)

die and response immediately

=item redirect($uri,status_code): Kossy::Response

=item render($file,$args): Kossy::Response

calls Text::Xslate->render makes response. template files are searching in root_dir/views directory

template syntax is Text::Xslate::Syntax::Kolon, can use Kossy::Connection object and fillinform block.

   ## template.tx
   : block form |  fillinform( $c.req ) -> {
   <head>
   <title><: $c.stash.title :></title>
   </head>
   <body>
   <form action="<: $c.req.uri_for('/post') :>">
   <input type="text" size="10" name="title" />
   <textarea name="body" rows="20" cols="90"></textarea>
   </form>
   </body>
   : }

also can use L<Text::Xslate::Bridge::TT2Like> and L<Number::Format> methods in your template

=item render_json($args): Kossy::Response

serializes arguments with JSON and makes response

This method escapes '<', '>', and '+' characters by "\uXXXX" form. Browser don't detects the JSON as HTML. And also this module outputs "X-Content-Type-Options: nosniff" header for IEs.

render_json have a JSON hijacking detection feature same as L<Amon2::Plugin::Web::JSON>. This returns "403 Forbidden" response if following pattern request.

=over 8

=item The request have 'Cookie' header.

=item The request doesn't have 'X-Requested-With' header.

=item The request contains /android/i string in 'User-Agent' header.

=item Request method is 'GET'

=back

=back

=head1 Kossy::Request

This class is child class of Plack::Request, decode query/body parameters automatically. Return value of $req->param(), $req->body_parameters, etc. is the decoded value.

=head2 OBJECT METHODS for Kossy::Request class

=over 4

=item uri_for($path,$args):String

build absolute URI with path and $args

  my $uri = $c->req->uri_for('/login',[ arg => 'Hello']);

=item validator($rule):Kossy::Validator::Result

validate parameters using L<Kossy::Validator>

  my $result = $c->req->validator([
    'q' => [['NOT_NULL','query must be defined']],
    'level' => {
        default => 'M',
        rule => [
            [['CHOICE',qw/L M Q H/],'invalid level char'],
        ],
    },
  ]);

  my $val = $result->valid('q');
  my $val = $result->valid('level');

=item body_parameters_raw

=item query_parameters_raw

=item parameters_raw

=item param_raw

These methods are the accessor to raw values. 'raw' means the value is not decoded.

=item body_parameters

Accessor to decoded body parameters. It's Hash::MultiValue object.

=item query_parameters

Accessor to decoded query parameters. It's Hash::MultiValue object.

=item json_parameters

Accessor to decoded JSON body parameters. It's B<NOT> Hash::MultiValue object.

    post '/api' => sub {
        my ($self, $c) = @_;
        my $foo = $c->req->json_parameters->{foo}; # bar
    };

    # requrest
    # $ua->requrest(
    #     HTTP::Request->new(
    #         "POST",
    #         "http://example.com/api",
    #         [ "Content-Type" => 'application/json', "Content-Length" => 13 ],
    #         '{"foo":"bar"}'
    #     )
    # );

NOTE: Not need to set C<kossy.request.parse_json_body> to 1

=back

=head1 Kossy::Response

This class is child class of Plack::Response

=head1 CUSTOMIZE

=over 4

=item X-Frame-Options

By default, Kossy outputs "X-Frame-Options: DENY". You can change this header

  get '/iframe' => sub {
      my ($self, $c) = @_;
      $c->res->header('X-Frame-Options','SAMEORIGIN');
      # or remove from response header
      # delete $c->res->headers->remove_header('X-Frame-Options');
      ..
  }

(Default: DENY)

=item kossy.request.parse_json_body

If enabled, Kossy will decode json in the request body that has "application/json" content header

  post '/api' => sub {
      my ($self, $c) = @_;
      $c->env->{'kossy.request.parse_json_body'} = 1;
      my val = $c->req->param('foo'); # bar
  }

  # requrest
  # $ua->requrest(
  #     HTTP::Request->new(
  #         "POST",
  #         "http://example.com/api",
  #         [ "Content-Type" => 'application/json', "Content-Length" => 13 ],
  #         '{"foo":"bar"}'
  #     )
  # );

=item $XSLATE_CACHE, $XSLATE_CACHE_DIR

Change xslate's cache level and cache directory.

  local $Kossy::XSLATE_CACHE = 2;
  local $Kossy::XSLATE_CACHE_DIR = $dir;
  my $app = MyApp::Web->psgi;

By default, $XSLATE_CACHE is 1, $XSLATE_CACHE_DIR is undef. use Xslate's default.

=item $SECURITY_HEADER

If disabled, Kossy does not set X-Frame-Options and X-XSS-Protection. enabled by default.

  local $Kossy::SECURITY_HEADER = 0;
  my $app = MyApp::Web->psgi;

Can not change $Kossy::SECURITY_HEADER in your WebApp. It's need to set at build time.

This is useful for the benchmark :-)

=item $JSON_SERIALIZER

changes the JSON serializer:

    use Cpanel::JSON::XS;
    use Cpanel::JSON::XS::Type;

    local $Kossy::JSON_SERIALIZER = Cpanel::JSON::XS->new()->allow_blessed(1)->convert_blessed(1)->ascii(0);

    get '/' => sub {
        my ($self, $c) = @_;
        return $c->render_json({ a => '234' }, { a => JSON_TYPE_INT });
    };

    my $app = __PACKAGE__->psgi;

=back

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo {at} gmail.comE<gt>

=head1 SEE ALSO

Kossy is small waf, that has only 400 lines code. so easy to reading framework code and customize it. Sinatra-ish router, build-in templating, validators and zero-configuration features are suitable for small application and rapid development.

L<Amon2::Lite>

L<Mojolicious::Lite>

L<Dancer>

L<Kossy::Validator>

=head1 LICENSE of HTTP::Headers::Fast

Kossy::Headers uses L<HTTP::Headers::Fast> code. Here is LICENSE of HTTP::Headers::Fast

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
